File: src/decrypt/go/decrypt.go
// Usage:
//   go run mosaic_decode.go "<ciphertext>"
//   go run mosaic_decode.go "<ciphertext>" "<key>"
// or build:
//   go build -o mosaic_decode mosaic_decode.go
//   ./mosaic_decode "<ciphertext>" "<key>"

package main

import (
	"bytes"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"strings"
	"unicode/utf8"
)

const (
	ALPHABET        = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?"
	NOISE_SET       = "abcdefghijklmnopqrstuvwxyz"
	TERM            = '~'
	BASE            = 47
	BLOCK_BYTES     = 5
	BLOCK_SYMBOLS   = 8
	CHECKSUM_PERIOD = 4
)

// build reverse map for a given alphabet string
func buildRev(alpha string) map[rune]int {
	rev := make(map[rune]int, len(alpha))
	for i, r := range alpha {
		rev[r] = i
	}
	return rev
}

var REV_BASE = buildRev(ALPHABET)

func rotationForBlock(blockIndex int) int {
	return ((blockIndex * 13) + 11) % BASE
}

func rotateAlpha(alpha string, rot int) string {
	n := len(alpha)
	if n == 0 {
		return ""
	}
	rot = ((rot % n) + n) % n
	return alpha[rot:] + alpha[:rot]
}

// base47 digits (most significant first) to 5 bytes (big-endian)
func base47DigitsTo5Bytes(digits []int) ([]byte, error) {
	if len(digits) != BLOCK_SYMBOLS {
		return nil, fmt.Errorf("expected %d digits, got %d", BLOCK_SYMBOLS, len(digits))
	}
	var val uint64 = 0
	for _, d := range digits {
		if d < 0 || d >= BASE {
			return nil, fmt.Errorf("digit out of range: %d", d)
		}
		val = val*BASE + uint64(d)
	}
	out := make([]byte, BLOCK_BYTES)
	for i := BLOCK_BYTES - 1; i >= 0; i-- {
		out[i] = byte(val & 0xFF)
		val >>= 8
	}
	return out, nil
}

func checksum47(blocksBytes [][]byte) int {
	var x byte = 0
	for _, b := range blocksBytes {
		for _, bb := range b {
			x ^= bb
		}
	}
	return int(x) % BASE
}

func isNoise(ch rune) bool {
	return strings.ContainsRune(NOISE_SET, ch)
}

func decodeMosaic(s string) ([]byte, error) {
	i := 0
	n := len(s)
	out := make([]byte, 0, 256)
	blockIndex := 0
	csWindows := make([][]byte, 0, CHECKSUM_PERIOD)
	revBase := REV_BASE

	for i < n {
		// skip whitespace
		for i < n && (s[i] == ' ' || s[i] == '\t' || s[i] == '\n' || s[i] == '\r' || s[i] == '\f' || s[i] == '\v') {
			i++
		}
		if i >= n {
			break
		}

		// trailer detection: "~~" + one base47 digit
		if i+2 < n && s[i] == byte(TERM) && s[i+1] == byte(TERM) {
			padChar := rune(s[i+2])
			index, ok := revBase[padChar]
			if !ok {
				return nil, fmt.Errorf("invalid trailer pad digit: %q", string(padChar))
			}
			padCount := index
			if padCount < 0 || padCount >= BLOCK_BYTES {
				return nil, fmt.Errorf("invalid pad_count: %d", padCount)
			}
			if padCount != 0 {
				if len(out) < padCount {
					return nil, errors.New("pad mismatch (output too small to trim)")
				}
				out = out[:len(out)-padCount]
			}
			i += 3
			if i != n {
				return nil, errors.New("extra data after trailer")
			}
			return out, nil
		}

		// prepare rotated alphabet for this block
		rot := rotationForBlock(blockIndex)
		rotated := rotateAlpha(ALPHABET, rot)
		revRot := buildRev(rotated)

		// read BLOCK_SYMBOLS digits, skipping noise characters
		digits := make([]int, 0, BLOCK_SYMBOLS)
		for k := 0; k < BLOCK_SYMBOLS; k++ {
			// skip noise characters
			for i < n && isNoise(rune(s[i])) {
				i++
			}
			if i >= n {
				return nil, errors.New("unexpected end of input while reading digits")
			}
			c := rune(s[i])
			if c == TERM {
				return nil, errors.New("unexpected terminator while expecting digit")
			}
			val, ok := revRot[c]
			if !ok {
				return nil, fmt.Errorf("invalid digit character for rotated alphabet: %q", string(c))
			}
			digits = append(digits, val)
			i++
		}

		// after digits, skip inserted noise then expect terminator
		for i < n && isNoise(rune(s[i])) {
			i++
		}
		if i >= n || s[i] != byte(TERM) {
			return nil, errors.New("missing block terminator after digits")
		}
		i++ // consume terminator

		// convert digits -> 5 bytes
		block5, err := base47DigitsTo5Bytes(digits)
		if err != nil {
			return nil, fmt.Errorf("converting digits to bytes: %w", err)
		}
		out = append(out, block5...)

		// add to checksum window
		csWindows = append(csWindows, block5)
		if len(csWindows) > CHECKSUM_PERIOD {
			csWindows = csWindows[len(csWindows)-CHECKSUM_PERIOD:]
		}

		blockIndex++

		// if checksum is due (window full)
		if len(csWindows) == CHECKSUM_PERIOD {
			// skip noise before checksum char
			for i < n && isNoise(rune(s[i])) {
				i++
			}
			if i >= n {
				return nil, errors.New("missing checksum character after block group")
			}
			chkChar := rune(s[i])
			i++
			got, ok := revBase[chkChar]
			if !ok {
				return nil, fmt.Errorf("invalid checksum char: %q", string(chkChar))
			}
			expect := checksum47(csWindows)
			if got != expect {
				return nil, fmt.Errorf("checksum mismatch: got %d expect %d", got, expect)
			}
			csWindows = csWindows[:0] // reset window
		}
	}

	return nil, errors.New("no trailer found; malformed input")
}

func xorWithKey(raw []byte, key []byte) []byte {
	if len(key) == 0 {
		return raw
	}
	out := make([]byte, len(raw))
	klen := len(key)
	for i := range raw {
		out[i] = raw[i] ^ key[i%klen]
	}
	return out
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s '<ciphertext>' [key]\n", os.Args[0])
		os.Exit(1)
	}
	s := os.Args[1]

	// match Python/C behavior: if key not provided, use fallback "default-key"
	var key string
	if len(os.Args) >= 3 {
		key = os.Args[2]
	} else {
		key = "default-key"
	}

	raw, err := decodeMosaic(s)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Decoding error:", err)
		os.Exit(2)
	}

	raw = xorWithKey(raw, []byte(key))

	fmt.Println("Decoded bytes (hex):", hex.EncodeToString(raw))

	if utf8.Valid(raw) {
		fmt.Println("Decoded text (utf-8):", string(raw))
	} else {
		var buf bytes.Buffer
		for _, b := range raw {
			if b >= 32 && b <= 126 {
				buf.WriteByte(b)
			} else {
				buf.WriteString(fmt.Sprintf("\\x%02x", b))
			}
		}
		fmt.Println("Decoded text:", buf.String())
	}
}

File: src/decrypt/java/Decrypt.java
import java.util.*;

public class Decrypt {
    static final String ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?";
    static final String NOISE_SET = "abcdefghijklmnopqrstuvwxyz";
    static final char TERM = '~';
    static final int BASE = 47;
    static final int BLOCK_BYTES = 5;
    static final int BLOCK_SYMBOLS = 8;
    static final int CHECKSUM_PERIOD = 4;

    // reverse lookup map
    static Map<Character, Integer> buildRev(String alpha) {
        Map<Character, Integer> rev = new HashMap<>();
        for (int i = 0; i < alpha.length(); i++) {
            rev.put(alpha.charAt(i), i);
        }
        return rev;
    }

    static final Map<Character, Integer> REV_BASE = buildRev(ALPHABET);

    static int rotationForBlock(int blockIndex) {
        return ((blockIndex * 13) + 11) % BASE;
    }

    static String rotateAlpha(String alpha, int rot) {
        int n = alpha.length();
        rot = ((rot % n) + n) % n;
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) {
            sb.append(alpha.charAt((i + rot) % n));
        }
        return sb.toString();
    }

    static byte[] base47DigitsTo5bytes(List<Integer> digits) {
        long val = 0;
        for (int d : digits) {
            val = val * BASE + d;
        }
        byte[] out = new byte[5];
        for (int i = 4; i >= 0; i--) {
            out[i] = (byte)(val & 0xFF);
            val >>= 8;
        }
        return out;
    }

    static int checksum47(List<byte[]> blocks) {
        int x = 0;
        for (byte[] b : blocks) {
            for (byte bb : b) {
                x ^= (bb & 0xFF);
            }
        }
        return x % BASE;
    }

    static byte[] decodeMosaic(String s) {
        int i = 0;
        int n = s.length();
        List<Byte> out = new ArrayList<>();
        int blockIndex = 0;
        List<byte[]> csWindows = new ArrayList<>();

        while (i < n) {
            // skip whitespace
            while (i < n && Character.isWhitespace(s.charAt(i))) i++;
            if (i >= n) break;

            // trailer detection
            if (i + 2 < n && s.charAt(i) == TERM && s.charAt(i + 1) == TERM) {
                char padChar = s.charAt(i + 2);
                if (!REV_BASE.containsKey(padChar)) {
                    throw new IllegalArgumentException("Invalid trailer pad digit: " + padChar);
                }
                int padCount = REV_BASE.get(padChar);
                if (padCount < 0 || padCount >= BLOCK_BYTES) {
                    throw new IllegalArgumentException("Invalid pad_count: " + padCount);
                }
                if (padCount > out.size()) {
                    throw new IllegalArgumentException("Pad mismatch");
                }
                for (int k = 0; k < padCount; k++) {
                    out.remove(out.size() - 1);
                }
                i += 3;
                if (i != n) {
                    throw new IllegalArgumentException("Extra data after trailer");
                }
                byte[] result = new byte[out.size()];
                for (int k = 0; k < out.size(); k++) result[k] = out.get(k);
                return result;
            }

            // rotated alphabet
            int rot = rotationForBlock(blockIndex);
            String rotated = rotateAlpha(ALPHABET, rot);
            Map<Character, Integer> revRot = buildRev(rotated);

            // read 8 digits
            List<Integer> digits = new ArrayList<>();
            for (int k = 0; k < BLOCK_SYMBOLS; k++) {
                while (i < n && NOISE_SET.indexOf(s.charAt(i)) != -1) i++;
                if (i >= n) throw new IllegalArgumentException("Unexpected end of input (digits)");
                char c = s.charAt(i);
                if (c == TERM) throw new IllegalArgumentException("Unexpected terminator");
                if (!revRot.containsKey(c)) {
                    throw new IllegalArgumentException("Invalid digit: " + c);
                }
                digits.add(revRot.get(c));
                i++;
            }

            while (i < n && NOISE_SET.indexOf(s.charAt(i)) != -1) i++;
            if (i >= n || s.charAt(i) != TERM) {
                throw new IllegalArgumentException("Missing block terminator");
            }
            i++;

            byte[] block5 = base47DigitsTo5bytes(digits);
            for (byte b : block5) out.add(b);

            csWindows.add(block5);
            if (csWindows.size() > CHECKSUM_PERIOD) {
                csWindows = csWindows.subList(csWindows.size() - CHECKSUM_PERIOD, csWindows.size());
            }

            blockIndex++;

            if (csWindows.size() == CHECKSUM_PERIOD) {
                while (i < n && NOISE_SET.indexOf(s.charAt(i)) != -1) i++;
                if (i >= n) throw new IllegalArgumentException("Missing checksum char");
                char chkChar = s.charAt(i++);
                if (!REV_BASE.containsKey(chkChar)) {
                    throw new IllegalArgumentException("Invalid checksum char: " + chkChar);
                }
                int got = REV_BASE.get(chkChar);
                int expect = checksum47(csWindows);
                if (got != expect) {
                    throw new IllegalArgumentException("Checksum mismatch: got " + got + " expect " + expect);
                }
                csWindows.clear();
            }
        }
        throw new IllegalArgumentException("No trailer found");
    }

    static byte[] xorWithKey(byte[] data, String key) {
        if (key == null || key.isEmpty()) return data;
        byte[] keyBytes = key.getBytes();
        byte[] out = new byte[data.length];
        for (int i = 0; i < data.length; i++) {
            out[i] = (byte)(data[i] ^ keyBytes[i % keyBytes.length]);
        }
        return out;
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java Decrypt <ciphertext> [key]");
            System.exit(1);
        }
        String ciphertext = args[0];
        String key = args.length >= 2 ? args[1] : "";

        try {
            byte[] raw = decodeMosaic(ciphertext);
            raw = xorWithKey(raw, key);

            // print hex
            StringBuilder hex = new StringBuilder();
            for (byte b : raw) {
                hex.append(String.format("%02x", b));
            }
            System.out.println("Decoded bytes (hex): " + hex);

            try {
                String text = new String(raw, "UTF-8");
                System.out.println("Decoded text (utf-8): " + text);
            } catch (Exception e) {
                System.out.println("Decoded text: (not valid UTF-8)");
            }
        } catch (Exception e) {
            System.err.println("Decoding error: " + e.getMessage());
            System.exit(2);
        }
    }
}
